File: ArraysAndHashing/Java/ContainsDuplicate/Duplicate.java
/*
 * Time complexity: O(n)
 *
 * Space complexity:
 * - Output: O(1) - only a boolean is returned
 * - Auxiliary: O(n) - due to the HashMap storing up to n elements (in worst case, all unique)
 * - Total: O(n)
 *
 * The logic:
 * We use a HashMap to track the frequency of each number as we iterate through the input.
 * If we see a duplicate, we return true immediately. Otherwise, we return false at the end.
 */

package Java.ContainsDuplicate;

import java.util.HashMap;

public class Duplicate {
  public static boolean hasDuplicate(int[] nums) {
    HashMap<Integer, Integer> freq = new HashMap<>();
    for (int num : nums) {
      freq.put(num, freq.getOrDefault(num, 0) + 1); // .getOrDef takes O(1) time
      if (freq.get(num) > 1)
        return true;
    }
    return false;
  }

  public static void main(String[] args) {
    int[] arr = { 1, 2, 3, 3 };
    System.out.println(hasDuplicate(arr));

    int[] arr2 = { 1, 2, 3, 4 };
    System.out.println(hasDuplicate(arr2));
  }
}

File: ArraysAndHashing/Python/ContainsDuplicate/duplicate.py
'''
 * Time complexity: O(n)
 *
 * Space complexity:
 * - Output: O(1) - only a boolean is returned
 * - Auxiliary: O(n) - due to the set storing up to n elements (if all unique)
 * - Total: O(n)
 *
 * The logic:
 * We use a set to track the numbers we have seen as we iterate through the input.
 * If we see a duplicate, we return True immediately. Otherwise, we return False at the end.
'''

from typing import List

class Solution:
    def hasDuplicate(self, nums: List[int]) -> bool:
      freq = set()
      for num in nums:
        if num in freq:
          return True
        freq.add(num)
      return False

if __name__ == "__main__":
    s = Solution()
    print(s.hasDuplicate([1, 2, 9, 15, 15]))
